update_copyright.d

```diff
@@ -0,0 +1,358 @@
+    static otherCopyrightRe = ctRegex!(`copyright.*[0-9][0-9]`, `i`);
+
+    // A regexp that matches one year.
+    static yearRe = ctRegex!(`[0-9?]+`);
+
+    // Matches part of a year or copyright holder.
+    static continuationRe = ctRegex!(rangesStr ~ `|` ~ nameStr);
+
+    static commentRe = ctRegex!(`#+|[*]+|;+|//+`);
+
+    // Convenience for passing around file/line number information.
+    struct FileLocation
+    {
+        string filename;
+        size_t linnum;
+
+        string toString()
+        {
+            import std.format;
+            return "%s(%d)".format(this.filename, this.linnum);
+        }
+    }
+
+    void processFile(string file)
+    {
+        // Looks like something we tried to create before.
+        if (file.endsWith(".tmp"))
+        {
+            remove(file);
+            return;
+        }
+
+        auto h = File(file, "rb");
+        auto lines = appender!string;
+        int errors = 0;
+        bool changed = false;
+
+        lines.reserve(h.size);
+
+        char[] orig_line = null;
+        char[] next_line = null;
+
+        auto loc = FileLocation(file, 0);
+
+        foreach (line; h.byLine)
+        {
+            loc.linnum++;
+            try
+            {
+                if (orig_line)
+                {
+                    auto continuation = this.stripContinuation(line);
+                    if (!continuation.matchAll(continuationRe))
+                    {
+                        // If the next line doesn't look like a proper continuation,
+                        // assume that what we've got is complete.
+                        lines.put(orig_line);
+                        lines.put('\n');
+                    }
+                    else
+                    {
+                        // Merge the lines for matching purposes.
+                        line = orig_line.stripRight() ~ `, ` ~ continuation;
+                    }
+                    orig_line = null;
+                }
+
+                auto match = line.matchAll(copyrightRe);
+                if (match)
+                {
+                    // If it looks like the copyright is incomplete, add the next line.
+                    if (!this.isComplete(match))
+                    {
+                        orig_line = line.dup;
+                        continue;
+                    }
+                    changed |= this.updateCopyright(loc, line, match, errors);
+                }
+                else if (line.matchAll(otherCopyrightRe))
+                {
+                    stderr.writeln(loc, ": unrecognised copyright: ", line.strip);
+                    //errors++;
+                }
+                lines.put(line);
+                lines.put('\n');
+            }
+            catch (Exception)
+            {
+                if (this.verbose)
+                    stderr.writeln(file, ": bad input file");
+                errors++;
+                break;
+            }
+        }
```

Comment:
I would move this big loop into a separate function, but that's subjective: http://wiki.c2.com/?LongFunctions